can/scripts/viewer.py

```diff
@@ -0,0 +1,623 @@
+    CANOPEN_LSS_RX:     {8: 'LSS_RX'},     # Layer Setting Services (LSS) receive
+}
+
+
+class CanViewer:
+
+    def __init__(self, stdscr, bus, data_structs, ignore_canopen, testing=False):
+        self.stdscr = stdscr
+        self.bus = bus
+        self.data_structs = data_structs
+        self.ignore_canopen = ignore_canopen
+
+        # Initialise the ID dictionary, start timestamp, scroll and variable for pausing the viewer
+        self.ids = {}
+        self.start_time = None
+        self.scroll = 0
+        self.paused = False
+
+        # Get the window dimensions - used for resizing the window
+        self.y, self.x = self.stdscr.getmaxyx()
+
+        # Do not wait for key inputs, disable the cursor and choose the background color automatically
+        self.stdscr.nodelay(True)
+        curses.curs_set(0)
+        curses.use_default_colors()
+
+        if not testing:  # pragma: no cover
+            self.run()
+
+    def run(self):
+        # Clear the terminal and draw the header
+        self.draw_header()
+
+        while 1:
+            # Do not read the CAN-Bus when in paused mode
+            if not self.paused:
+                # Read the CAN-Bus and draw it in the terminal window
+                msg = self.bus.recv(timeout=0)
+                if msg is not None:
+                    self.draw_can_bus_message(msg)
+
+            # Read the terminal input
+            key = self.stdscr.getch()
+
+            # Stop program if the user presses ESC or 'q'
+            if key == KEY_ESC or key == ord('q'):
+                break
+
+            # Clear by pressing 'c'
+            elif key == ord('c'):
+                self.ids = {}
+                self.start_time = None
+                self.scroll = 0
+                self.draw_header()
+
+            # Pause by pressing space
+            elif key == KEY_SPACE:
+                self.paused = not self.paused
+
+            # Scroll by pressing up/down
+            elif key == curses.KEY_UP:
+                # Limit scrolling, so the user do not scroll passed the header
+                if self.scroll > 0:
+                    self.scroll -= 1
+                    self.redraw_screen()
+            elif key == curses.KEY_DOWN:
+                # Limit scrolling, so the maximum scrolling position is one below the last line
+                if self.scroll <= len(self.ids) - self.y + 1:
+                    self.scroll += 1
+                    self.redraw_screen()
+
+            # Check if screen was resized
+            resized = curses.is_term_resized(self.y, self.x)
+            if resized is True:
+                self.y, self.x = self.stdscr.getmaxyx()
+                if hasattr(curses, 'resizeterm'):  # pragma: no cover
+                    curses.resizeterm(self.y, self.x)
+                self.redraw_screen()
+
+            # Sleep 0.1 ms, so the application does not use 100 % of the CPU resources
+            time.sleep(.1 / 1000.)
```

Comment:
I think it's better to put a longer timeout in the `recv` method instead. By sleeping between each message you risk losing messages eventually. A timeout will only allow the program to rest when the receive queue has been cleared.

Reply:
Done.